### PR TITLE
coverage: Allow each coverage statement to have multiple code regions

### PR DESCRIPTION
**Pull request by Zalathar**
The original implementation of coverage instrumentation was built around the assumption that a coverage counter/expression would be associated with *up to one* code region. When it was discovered that *multiple* regions would sometimes need to share a counter, a workaround was found: for the remaining regions, the instrumentor would create a fresh expression that adds zero  to the existing counter/expression.

That got the job done, but resulted in some awkward code, and produces unnecessarily complicated coverage maps in the final binary.

---

This PR removes that tension by changing `StatementKind::Coverage`'s code region field from `Option<CodeRegion>` to `Vec<CodeRegion>`.

The changes on the codegen side are fairly straightforward. As long as each `CoverageKind::Counter` only injects one `llvm.instrprof.increment`, the rest of coverage codegen is happy to handle multiple regions mapped to the same counter/expression, with only minor option-to-vec adjustments.

On the instrumentor/mir-transform side, we can get rid of the code that creates extra (x + 0) expressions. Instead we gather all of the code regions associated with a single BCB, and inject them all into one coverage statement.

---

There are several patches here but they can be divided in to three phases:
- Preparatory work
- Actually switching over to multiple regions per coverage statement
- Cleaning up

So viewing the patches individually may be easier.

### CODE REVIEW

**Comment by Zalathar** [comment]
(Cargo.toml changes are part of #114843 and won't remain in the final version of this PR.)

File: compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs [hunk 1]
```diff
@@ -110,25 +110,15 @@ impl<'tcx> CoverageInfoBuilderMethods<'tcx> for Builder<'_, '_, 'tcx> {
...
                 func_coverage.set_function_source_hash(function_source_hash);
-
-                if let Some(code_region) = code_region {
-                    // Note: Some counters do not have code regions, but may still be referenced
-                    // from expressions. In that case, don't add the counter to the coverage map,
-                    // but do inject the counter intrinsic.
-                    debug!(
-                        "adding counter to coverage_map: instance={:?}, id={:?}, region={:?}",
-                        instance, id, code_region,
-                    );
```
Comment (Zalathar) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs]: These `debug!` invocations have been replaced with `#[instrument(...)]` tracing on the methods being called.

File: compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs [hunk 2]
```diff
@@ -110,25 +110,15 @@ impl<'tcx> CoverageInfoBuilderMethods<'tcx> for Builder<'_, '_, 'tcx> {
...
-                    func_coverage.add_counter(id, code_region);
-                }
+                func_coverage.add_counter(id, code_regions);
```
Comment (Zalathar) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs:121]: We now call `add_counter` unconditionally here, and leave the callee to decide how to handle counters with zero code regions.

File: compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs [hunk 3]
```diff
@@ -156,156 +223,66 @@ impl<'tcx> FunctionCoverage<'tcx> {
...
-        self.counters.iter_enumerated().filter_map(|(index, entry)| {
-            // Option::map() will return None to filter out missing counters. This may happen
-            // if, for example, a MIR-instrumented counter is removed during an optimization.
-            entry.as_ref().map(|region| (Counter::counter_value_reference(index), region))
-        })
+        self.counters
+            .iter_enumerated()
+            // Filter out counter IDs that we never saw during MIR traversal.
+            // This can happen if a counter was optimized out by MIR transforms.
+            // (Its regions will have been remapped to zero instead.)
```
Comment (Swatinem) [inline compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:230]:
```suggestion
            // (It was replaced by an `Unreachable` mapping)
```

File: tests/coverage-map/status-quo/async.cov-map [hunk 4]
```diff
@@ -148,14 +146,14 @@ Number of file 0 mappings: 1
...
+Raw bytes (69): 0x[01, 01, 00, 0d, 01, 15, 17, 01, 0c, 00, 02, 09, 00, 0a, 00, 00, 0e, 00, 11, 00, 00, 12, 00, 17, 00, 00, 1b, 00, 1c, 00, 00, 20, 00, 22, 00, 01, 09, 00, 0a, 00, 00, 0e, 00, 11, 00, 00, 12, 00, 17, 00, 00, 1b, 00, 1c, 00, 00, 20, 00, 22, 00, 01, 0e, 00, 10, 00, 02, 01, 00, 02]
 Number of files: 1
 - file 0 => global file 1
 Number of expressions: 0
 Number of file 0 mappings: 13
-- Code(Zero) at (prev + 21, 23) to (start + 1, 12)
+- Code(Counter(0)) at (prev + 21, 23) to (start + 1, 12)
 - Code(Zero) at (prev + 2, 9) to (start + 0, 10)
-- Code(Counter(0)) at (prev + 0, 14) to (start + 0, 17)
+- Code(Zero) at (prev + 0, 14) to (start + 0, 17)
```
[the remark this answers is not in the document]
Comment (Zalathar) [reply tests/coverage-map/status-quo/async.cov-map]: There's some existing code in `add_unused_function_coverage` that ensures each unused function's coverage has at least one `Counter` mapping.

(I'm not convinced it's necessary, but I don't want to do the homework of proving that it's *not* necessary, so for now I've left it as-is.)

**Comment by davidtwco** [comment]
There's still lots of discussion and changes happening here, use `@rustbot ready` when you've got something you're looking to have merged.

**Comment by Zalathar** [comment]
I've folded some `spans` changes into this [(diff)](https://github.com/rust-lang/rust/compare/2037b9860bccb4c3b704d994bf5735e43f5b3700..5ede04f1ba6cd2b6c0cc92f455aca6827ef27376), because they make it easier to get all of the coverage spans associated with a BCB.

**Comment by Zalathar** [comment]
#114399 has landed, so I'm going to give this PR another round of polish and then flag it for review.

**Comment by Zalathar** [comment]
@davidtwco All dependencies have landed, so this is now ready for review.
@rustbot ready

**Review by Swatinem — APPROVED** [review]

**Comment by Zalathar** [comment]
(I've been rebasing this for the benefit of other PRs/patches that rely on it.)

**Review by davidtwco — APPROVED** [review]
Apologies for the delay in getting to this, looks good to me.

**Pull request merged**